pulser/simulation/simulation.py

```diff
@@ -118,24 +162,42 @@ def _set_param_from_config(self) -> None:
         Called every time a new configuration is loaded in order to update
         Simulation parameters.
         """
-        self._prepare_bad_atoms()
+        self._update_noise()
         # update samples with potential noise settings
-        self._build_hamiltonian_from_seq()
+        self._extract_samples()
         if 'dephasing' in self.config.noise:
             self._init_dephasing()
 
-    def _prepare_bad_atoms(self) -> None:
-        """Chooses atoms that will be badly prepared.
+    def _update_noise(self) -> None:
+        """Updates noise random parameters.
 
-        If no SPAM noise, simply sets all atoms as well-prepared.
+        Used at the start of each run. If SPAM isn't in chosen noises, all
+            atoms are set to be correctly prepared.
         """
-        # Tag True if atom is badly prepared
-        self._bad_atoms: dict[Union[str, int], bool] = {
-            qid: False for qid in self._qid_index}
         if 'SPAM' in self.config.noise:
-            dist = (np.random.uniform(size=len(self._qid_index)) <
-                    self.config.spam_dict['eta'])
-            self._bad_atoms = dict(zip(self._qid_index, dist))
+            self._prepare_bad_atoms()
```

Comment:
I don't know if this is negligibly fast, but should you do this if `eta` == 0?

Reply:
I guess not, but I doubt it speeds up anything. Will add a check anyway.